browser/caliban.py

```diff
@@ -473,6 +475,111 @@ def action_threshold(self, y1, x1, y2, x2, label):
         if np.any(np.isin(self.frame[..., self.feature], label)):
             self.add_cell_info(add_label=label, frame=self.frame_id)
 
+    def action_active_contour(self, label):
+        label_img = np.copy(self.frame[..., self.feature])
+
+        # get centroid of selected label
+        props = regionprops(np.where(label_img == label, label, 0))[0]
+
+        # make bounding box size to encompass some background
+        box_height = props['bbox'][2] - props['bbox'][0]
+        y1 = max(0, props['bbox'][0] - box_height // 2)
+        y2 = min(self.project.height, props['bbox'][2] + box_height // 2)
+
+        box_width = props['bbox'][3] - props['bbox'][1]
+        x1 = max(0, props['bbox'][1] - box_width // 2)
+        x2 = min(self.project.width, props['bbox'][3] + box_width // 2)
+
+        # relevant region of label image to work on
+        label_img = label_img[y1:y2, x1:x2]
+
+        # use existing label as initial level set for contour calculations
+        level_set = np.where(label_img == label, 1, 0)
+
+        # normalize input 2D frame data values to range [0.0, 1.0]
+        adjusted_raw_frame = Normalize()(self.raw_frame[..., self.channel])
+        predict_area = adjusted_raw_frame[y1:y2, x1:x2]
+        # import pdb; pdb.set_trace()
+
+        # returns 1 where label is predicted to be based on contouring, 0 background
+        contoured = morphological_chan_vese(predict_area, 100, init_level_set=level_set)
+
+        # contoured area should get original label value
+        contoured_label = np.where(contoured, label, 0)
+        # contours tend to fit very tightly, a small expansion here works well
+        contoured_label = dilation(contoured_label, disk(3))
+
+        # don't want to leave the original (un-contoured) label in the image
+        label_img = np.where(label_img == label, 0, label_img)
+
+        # never overwrite other labels with new contoured label
+        safe_overlay = np.where(label_img == 0, contoured_label, label_img)
+
+        # label must be present in safe_overlay for this to be a valid contour result
+        # very few pixels of contoured label indicate contour prediction not worth keeping
+        pixel_count = np.count_nonzero(safe_overlay == label)
+        if pixel_count < 20:
+            safe_overlay = np.copy(self.frame[y1:y2, x1:x2, self.feature])
+
+        # put it back in the full image so can use centroid coords for post-contour cleanup
+        full_frame = np.copy(self.frame[..., self.feature])
+        full_frame[y1:y2, x1:x2] = safe_overlay
+
+        # avoid automated label cleanup if the centroid (flood seed point) is of background
+        # TODO: perhaps this should be "if (seed point value) != label" instead
```

Comment:
Let's make this change, redeploy, and do a quick sanity test of functionality. This would only make a difference in some very weird edge cases but it ought to be quick and easy to change now.